{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get data shapes, i.e. numbers of features and samples for each drug in only targe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General imports\n",
    "import multiprocessing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import time\n",
    "import sys\n",
    "import dill\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import pearsonr\n",
    "import collections\n",
    "import os\n",
    "\n",
    "# Sklearn imports\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn import model_selection\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn import metrics\n",
    "from sklearn.dummy import DummyRegressor\n",
    "from sklearn.linear_model import Lasso, ElasticNet\n",
    "from stability_selection import StabilitySelection\n",
    "\n",
    "# Add directory to sys.path in order to import custom modules from there.\n",
    "sys.path.insert(0, \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Projects/Created Modules\")\n",
    "from gdsc_projects_module import DrugWithDrugBank, Experiment, Modeling, ModelingResults\n",
    "\n",
    "# Initialize proper file pathways\n",
    "drug_annotations = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Drug annotations/Screened_Compounds-March_27th_2018.xlsx\"\n",
    "cell_line_list = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Cell line list (directly from website)/Cell_listThu Aug 16 22_06_49 2018.csv\"\n",
    "gene_expr = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Gene expression/sanger1018_brainarray_ensemblgene_rma-March_2nd_2017.txt\"\n",
    "cnv1 = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Copy number variations/cnv_binary_1.csv\"\n",
    "cnv2 = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Copy number variations/PANCANCER_Genetic_feature_cna_Mon Aug  6 16_18_51 2018 (kopia).csv\"\n",
    "coding_variants = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Mutation calls/PANCANCER_Genetic_feature_variant_Mon Aug  6 15_45_44 2018.csv\"\n",
    "drug_response = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Sensitivity profiles/v17.3_fitted_dose_response-March_27th_2018.xlsx\"\n",
    "\n",
    "# Load dictionary with targets derived from DrugBank\n",
    "drugbank_targets = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/DrugBank/Created data/drugbank_map_drug_to_targets.p\"\n",
    "\n",
    "# Filepath to gene expression signatures provided by Merck\n",
    "signatures = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Created data/Merck Gene Expression Signatures/Data/SignatureScores_GDSC-cellLines_2018-09-27.tsv\"\n",
    "\n",
    "# Call loading function from DrugWithDrugBank class\n",
    "(drug_annotations_df, cell_lines_list_df, gene_expression_df, cnv_binary_df, \n",
    " coding_variants_df, drug_response_df, map_drugs_to_drugbank_targets) = DrugWithDrugBank.load_data(\n",
    "    drug_annotations, cell_line_list, gene_expr, \n",
    "    cnv1, cnv2, coding_variants, drug_response, drugbank_targets)\n",
    "\n",
    "# Load gene expression signatures\n",
    "signatures_df = pd.read_table(signatures)\n",
    "\n",
    "# Load helper dict for extraction of CNV data\n",
    "filepath = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Data/Original Data/Genomics of Drug Sensitivity in Cancer/Original GDSC Data/Copy number variations/Created data/\"\n",
    "with open(filepath + \"map_cl_id_and_genetic_feature_to_mutation_status.pkl\", \"rb\") as f:\n",
    "    map_from_cl_id_and_genetic_feature_to_mutation_status = dill.load(f)\n",
    "\n",
    "# Load gene mappings\n",
    "filepath1 = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Projects/GDSC - Prediction only with data related to nominal drug targets (minimal approach)/Created data/mapping_from_ensembl_id_to_hgnc_symbol.p\"\n",
    "filepath2 = \"/home/kkoras/Documents/Projects/Doktorat - Modelling drug efficacy in cancer/Projects/GDSC - Prediction only with data related to nominal drug targets (minimal approach)/Created data/mapping_from_hgnc_symbol_to_ensembl_id.p\"\n",
    "DrugWithDrugBank.load_mappings(filepath2, filepath1)   # Initialize class variables\n",
    "\n",
    "# Print shapes of created DataFrames\n",
    "print(\"Loading summary:\")\n",
    "print(\"Drug annotations:\", drug_annotations_df.shape)\n",
    "print(\"Cell line list\", cell_lines_list_df.shape)\n",
    "print(\"Gene expression\", gene_expression_df.shape)\n",
    "print(\"CNV binary:\", cnv_binary_df.shape)\n",
    "print(\"Coding variants:\", coding_variants_df.shape)\n",
    "print(\"Drug response:\", drug_response_df.shape)\n",
    "print(\"DrugBank mapping (number of matched drugs):\", len(map_drugs_to_drugbank_targets))\n",
    "print(\"Gene expression signatures:\", signatures_df.shape)\n",
    "print(\"Number of entries in mapping from cell line and cnv genetic feature to mutation status:\",\n",
    "     len(map_from_cl_id_and_genetic_feature_to_mutation_status))\n",
    "\n",
    "# Create drug objects\n",
    "drugs = DrugWithDrugBank.create_drugs(drug_annotations_df, map_drugs_to_drugbank_targets)\n",
    "print(len(drugs))\n",
    "\n",
    "# Set up data types we want to include in our input for each drug\n",
    "data_types = [\"CNV\", \"mutation\", \"expression\", \"tissue\"]\n",
    "# Create input data\n",
    "Experiment.create_input_for_each_drug(drugs, drug_response_df, data_combination=data_types, \n",
    "                                     gene_expression_df=gene_expression_df, \n",
    "                                     cnv_binary_df=cnv_binary_df,\n",
    "                                     map_cl_id_and_feature_to_status=map_from_cl_id_and_genetic_feature_to_mutation_status,\n",
    "                                     cell_lines_list_df=cell_lines_list_df,\n",
    "                                     coding_variants_df=coding_variants_df,\n",
    "                                     feat_threshold=16,\n",
    "                                     log=True)\n",
    "\n",
    "# Hyperparameter space to search on\n",
    "param_grid = {\"estimator__alpha\": [0.0001, 0.001, 0.01, 0.1, 1., 5., 10., 30., 50., 100.],\n",
    "              \"estimator__l1_ratio\": [0.0, 0.1, 0.25, 0.5, 0.75, 0.9, 1.]}\n",
    "# Compute the number of all possible combinations\n",
    "all_combinations = 1\n",
    "for param in param_grid:\n",
    "    all_combinations *= len(param_grid[param])\n",
    "print(all_combinations)\n",
    "\n",
    "enet_seeds = [22, 37, 44, 55, 78]\n",
    "split_seeds = [11, 37, 52, 71, 98]\n",
    "\n",
    "exp = Modeling(name=\"Only targets - modeling with ENet over few data splits\",\n",
    "              param_grid=param_grid,\n",
    "              estimator_seeds=enet_seeds,\n",
    "              split_seeds=split_seeds,\n",
    "              n_combinations=30,\n",
    "              kfolds=3,\n",
    "              max_iter=2000,\n",
    "              tuning_jobs=4)\n",
    "\n",
    "# Initialize new ModelingResults object\n",
    "exp_results = ModelingResults(exp)\n",
    "print(exp_results.kfolds, exp_results.tuning_jobs, exp_results.scoring, exp_results.max_iter)\n",
    "\n",
    "# Load previously computed results\n",
    "# filename = \"\"\n",
    "# with open(\"../Created data/Results/\" + filename, \"rb\") as f:\n",
    "#     exp_results = dill.load(f)\n",
    "\n",
    "# Get rid of warnings\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "drug_counter = 0\n",
    "log = True   # Controls verbosity during iterating over drugs\n",
    "\n",
    "# Enter the loop over drugs\n",
    "for drug_id in drugs:\n",
    "    drug = drugs[drug_id]   # Current Drug object\n",
    "    data = drug.full_data  # Extract input data (should be previously computed)\n",
    "    if data.shape[0] == 0:   # Check if data exists, if not, skip the drug\n",
    "        continue\n",
    "    if data.shape[1] < 16:    # That means that data has only features related to tissue\n",
    "        continue           # so also skip this case\n",
    "        \n",
    "    if log:\n",
    "        print(drug.name, data.shape)\n",
    "        \n",
    "    # Extract features and labels\n",
    "    y = data[\"AUC\"]\n",
    "    X = data.drop([\"cell_line_id\", \"AUC\"], axis=1)\n",
    "    X.shape[0] == y.shape[0]\n",
    "        \n",
    "    # Add data shapes to corresponding dictionary field in ModelingResults\n",
    "    exp_results.data_shapes[(drug.name, drug_id)] = X.shape\n",
    "    \n",
    "    # Compute the results\n",
    "    (test_results_for_splits, cv_results_for_splits, \n",
    "     best_parameters_for_splits, \n",
    "     dummy_for_splits, tuning_seeds_for_splits) = exp.enet_model_over_data_splits(X, y, verbose=1, log=True)\n",
    "    \n",
    "    # Put results into appropriate fields of ModelingResults object\n",
    "    exp_results.performance_dict[(drug.name, drug_id)] = test_results_for_splits\n",
    "    exp_results.dummy_performance_dict[(drug.name, drug_id)] = dummy_for_splits\n",
    "    exp_results.best_params_dict[(drug.name, drug_id)] = best_parameters_for_splits\n",
    "    exp_results.tuning_seeds_dict[(drug.name, drug_id)] = tuning_seeds_for_splits\n",
    "    exp_results.cv_results_dict[(drug.name, drug_id)] = cv_results_for_splits\n",
    "    \n",
    "    drug_counter +=1 \n",
    "    print(drug_counter, \"drugs done\")\n",
    "    print()\n",
    "    print(\"*\" * 50)\n",
    "    print()\n",
    "\n",
    "print()\n",
    "print(\"SCRIPT FINISHED, ALL DRUGS DONE\")\n",
    "print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (modeling_drug_efficacy)",
   "language": "python",
   "name": "modeling_drug_efficacy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
